File: windows/src/GoalMaker.Core/Account/EmailAddress.cs
using System.Text.RegularExpressions;

namespace GoalMaker.Core.Account;

/// <summary>A trimmed, plausibly valid email address. Supabase Auth does the real check.</summary>
public sealed partial record EmailAddress
{
    private EmailAddress(string value) => Value = value;

    public string Value { get; }

    public static EmailAddress? Parse(string text)
    {
        var trimmed = text.Trim();
        return trimmed.Length <= 254 && Pattern().IsMatch(trimmed) ? new EmailAddress(trimmed) : null;
    }

    public override string ToString() => Value;

    [GeneratedRegex(@"^[^@\s]+@[^@\s]+\.[^@\s]+\z", RegexOptions.CultureInvariant)]
    private static partial Regex Pattern();
}

File: android/app/src/main/kotlin/com/goalmaker/app/domain/account/EmailAddress.kt
package com.goalmaker.app.domain.account

/** A trimmed, plausibly valid email address. Supabase Auth does the real check. */
@JvmInline
value class EmailAddress private constructor(val value: String) {
    companion object {
        private val pattern = Regex("^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$")

        fun parse(text: String): EmailAddress? =
            text.trim().takeIf { it.length <= 254 && pattern.matches(it) }?.let(::EmailAddress)
    }
}
